src/libs/tools/lighttable.c

```diff
@@ -242,26 +291,65 @@ static void _lib_lighttable_key_accel_toggle_culling_dynamic_mode(dt_action_t *a
   dt_lib_module_t *self = darktable.view_manager->proxy.lighttable.module;
   dt_lib_tool_lighttable_t *d = (dt_lib_tool_lighttable_t *)self->data;
 
-  // if we are already in any culling layout, we return to the base layout
-  if(d->layout != DT_LIGHTTABLE_LAYOUT_CULLING && d->layout != DT_LIGHTTABLE_LAYOUT_CULLING_DYNAMIC)
-    _lib_lighttable_set_layout(self, DT_LIGHTTABLE_LAYOUT_CULLING_DYNAMIC);
-  else
+  // if we are already in culling dynamic layout, we return to the base layout
+  if(d->layout == DT_LIGHTTABLE_LAYOUT_CULLING_DYNAMIC)
     _lib_lighttable_set_layout(self, d->base_layout);
+  else
+    _lib_lighttable_set_layout(self, DT_LIGHTTABLE_LAYOUT_CULLING_DYNAMIC);
 
   dt_control_queue_redraw_center();
 }
 
-static void _lib_lighttable_key_accel_toggle_culling_mode(dt_action_t *action)
+static void _lib_lighttable_key_accel_toggle_culling_movement_restriction(dt_action_t *action)
 {
   dt_lib_module_t *self = darktable.view_manager->proxy.lighttable.module;
   dt_lib_tool_lighttable_t *d = (dt_lib_tool_lighttable_t *)self->data;
 
-  // if we are already in any culling layout, we return to the base layout
-  if(d->layout != DT_LIGHTTABLE_LAYOUT_CULLING && d->layout != DT_LIGHTTABLE_LAYOUT_CULLING_DYNAMIC)
-    _lib_lighttable_set_layout(self, DT_LIGHTTABLE_LAYOUT_CULLING);
+  // if we are already in unrestricted culling mode, restrict movement
+  // otherwise enter unrestricted mode first
+  if(d->layout == DT_LIGHTTABLE_LAYOUT_CULLING)
+    _lib_lighttable_set_layout(self, DT_LIGHTTABLE_LAYOUT_CULLING_RESTRICTED);
   else
-    _lib_lighttable_set_layout(self, d->base_layout);
+    _lib_lighttable_set_layout(self, DT_LIGHTTABLE_LAYOUT_CULLING);
+  dt_control_queue_redraw_center();
+}
 
+static void _lib_lighttable_key_accel_toggle_culling_restricted_mode(dt_action_t *action)
+{
+  dt_lib_module_t *self = darktable.view_manager->proxy.lighttable.module;
+  dt_lib_tool_lighttable_t *d = (dt_lib_tool_lighttable_t *)self->data;
+
+  // if we are already in restricted culling mode, leave the mode
+  if(d->layout == DT_LIGHTTABLE_LAYOUT_CULLING || d->layout == DT_LIGHTTABLE_LAYOUT_CULLING_RESTRICTED)
+    _lib_lighttable_set_layout(self, d->base_layout);
+  else // otherwise first check, what the selection count is
+  {
+    int sel_count = 0;
+    sqlite3_stmt *stmt;
+    // clang-format off
+    DT_DEBUG_SQLITE3_PREPARE_V2(dt_database_get(darktable.db),
+                                "SELECT count(*) "
+                                "FROM memory.collected_images AS col, main.selected_images as sel "
+                                "WHERE col.imgid=sel.imgid",
+                                -1, &stmt, NULL);
+    // clang-format on
+    if(sqlite3_step(stmt) == SQLITE_ROW) sel_count = sqlite3_column_int(stmt, 0);
+    sqlite3_finalize(stmt);
```

Comment:
if we replace all that by : `int sel_count = dt_act_on_get_images_nb(TRUE, FALSE);` that will : 
- simplify the code
- avoid current behavior change with current algo
- be future proof once the new algo is done as it will propose another way to compute `dt_act_on_get_images_nb`. esp in the "selection driven algo", it will return the nb of selected images